builder/dockerfile/builder.go

```diff
@@ -177,62 +176,54 @@ func emitImageID(aux *streamformatter.AuxFormatter, state *dispatchState) error
 	}
 	return aux.Emit(types.BuildResult{ID: state.imageID})
 }
-
-func (b *Builder) dispatchDockerfileWithCancellation(dockerfile *parser.Result, source builder.Source) (*dispatchState, error) {
-	shlex := NewShellLex(dockerfile.EscapeToken)
-	state := newDispatchState()
-	total := len(dockerfile.AST.Children)
-	var err error
-	for i, n := range dockerfile.AST.Children {
-		select {
-		case <-b.clientCtx.Done():
-			logrus.Debug("Builder: build cancelled!")
-			fmt.Fprint(b.Stdout, "Build cancelled")
-			buildsFailed.WithValues(metricsBuildCanceled).Inc()
-			return nil, errors.New("Build cancelled")
-		default:
-			// Not cancelled yet, keep going...
+func (b *Builder) dispatchDockerfileWithCancellation(parseResult typedcommand.BuildableStages, metaArgs []typedcommand.ArgCommand, escapeToken rune, source builder.Source) (*dispatchState, error) {
+	var dispatcher *dispatcher
+	// metaargs dispatcher
+	dispatcher = newDispatcher(b, escapeToken, source)
+	for _, meta := range metaArgs {
+		if err := dispatcher.dispatch(&meta); err != nil {
+			return nil, err
 		}
+	}
 
-		// If this is a FROM and we have a previous image then
-		// emit an aux message for that image since it is the
-		// end of the previous stage
-		if n.Value == command.From {
-			if err := emitImageID(b.Aux, state); err != nil {
-				return nil, err
+	for _, stage := range parseResult {
+		dispatcher = newDispatcher(b, escapeToken, source)
+		for i, cmd := range stage.Commands {
+			select {
+			case <-b.clientCtx.Done():
+				logrus.Debug("Builder: build cancelled!")
+				fmt.Fprint(b.Stdout, "Build cancelled")
+				buildsFailed.WithValues(metricsBuildCanceled).Inc()
+				return nil, errors.New("Build cancelled")
+			default:
+				// Not cancelled yet, keep going...
 			}
-		}
+			sourceCode := "..."
+			if src, ok := cmd.(typedcommand.WithSourceCode); ok {
+				sourceCode = src.SourceCode()
+			}
+			if len(parseResult) > 1 {
+				fmt.Fprintf(b.Stdout, "Stage %v: %v / %v %v", stage.Name, i+1, len(stage.Commands), sourceCode)
```

Comment:
Maybe avoid doing new logging output for now as it will likely change more in the future. No issue if it changes a bit just because the logic is different.